File: Cores/OSOL.Extremum.Core.DotNet/CodeFeatures/FunctionParser/Trees/Tree.cs
using System;
using OSOL.Extremum.Core.DotNet.Vectors;

namespace OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.Trees
{
    public abstract class Tree<T>
    {
        public abstract T Calculate(VectorObject<T> v);
    }

    public abstract class BinaryOpTree<T> : Tree<T>
    {
        public Tree<T> Left, Right;
        public Func<T, T, T> Op;

        public override T Calculate(VectorObject<T> v) => Op(Left.Calculate(v), Right.Calculate(v));
    }
    
    public abstract class UnaryOpTree<T> : Tree<T>
    {
        public Tree<T> SubTree;
        public Func<T, T> Op;

        public override T Calculate(VectorObject<T> v) => Op(SubTree.Calculate(v));
    }

    
}

File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/CodeFeatures/FunctionParser/Trees/Tree.scala
package OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Trees

import OSOL.Extremum.Core.Scala.Vectors.VectorObject

abstract class Tree[T] {
  def calculate(v: VectorObject[T]): T
}

object Tree {
  abstract class BinaryOpTree[T](left: Tree[T], right: Tree[T], op: (T, T) => T) extends Tree[T] {
    final override def calculate(v: VectorObject[T]): T = op(left.calculate(v), right.calculate(v))
  }

  abstract class UnaryOpTree[T](subTree: Tree[T], op: T => T) extends Tree[T] {
    final override def calculate(v: VectorObject[T]): T = op(subTree.calculate(v))
  }
}